src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -6182,7 +6267,15 @@ private static bool DoMultipleAddressConnectCallback(object result, MultipleAddr
                 {
                     try
                     {
-                        context.socket.EndConnect((IAsyncResult)result);
+                        if (SocketPal.SupportsMultipleConnectAttempts || context.isUserConnectAttempt)
+                        {
+                            context.socket.EndConnect((IAsyncResult)result);
+                        }
+                        else
+                        {
+                            Debug.Assert(context.lastAttemptSocket != null);
+                            context.lastAttemptSocket.EndConnect((IAsyncResult)result);
+                        }
```

Comment:
If we're not in the SupportsMultipleConnectAttempts mode, is lastAttemptSocket just null?  What if you were to make lastAttemptSocket always the thing you call EndConnect on (same goes for the other similar blocks) by making sure that even in the SupportsMultipleConnectAttempts mode lastAttemptSocket = context.socket.  Could you then simplify a bunch of these branches by just always doing the operation on lastAttemptSocket?